_drafts/restless-restfull.md

```diff
@@ -0,0 +1,456 @@
+since each interface has its own set of operations that is documented,
+there is less restrictions (also more space to inconsistencies in naming
+and other things, tradeoffs, no free lunch, etc).
+
+So REST and its more known implementation on top of HTTP has a fixed
+set of methods with a specific semantic for each one.
+
+Before going deeper on the semantics of the methods, you must ask yourself,
+why it is so important to use the correct semantic of each method ?
+
+The first that always came to my mind is that consistency makes your API
+easier to understand, you can use intuition to infer how to do stuff instead
+of always reading docs. Being able to use intuition to do something on an API
+is a sign of a great API to me. But when we are thinking about distributed
+systems there are other important things to consider.
+
+The most important one, that always bites us in the ass, is side effects.
+Our inability to properly handle side effects is what makes functional
+programming so attractive, we suck at it.
+
+So how does HTTP specifies state changes ? Lets take a look on the
+[spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html):
+
+```
+In particular, the convention has been established that the GET and HEAD
+methods SHOULD NOT have the significance of taking an action
+other than retrieval.
+
+These methods ought to be considered "safe". This allows user agents
+to represent other methods, such as POST, PUT and DELETE, in a special way,
+so that the user is made aware of the fact that a possibly unsafe
+action is being requested.
+
+Naturally, it is not possible to ensure that the server does not
+generate side-effects as a result of performing a GET request; in fact,
+some dynamic resources consider that a feature. The important distinction
+here is that the user did not request the side-effects,
+so therefore cannot be held accountable for them.
+```
+
+The advantage of being explicit about state changes is a more obvious one.
+If the system is built properly and some code is only performing GET's you
+can be sure that it is not the culprit for some state change (it also has
+applications on caching, which is pretty important to scale in some contexts).
+
+There are more subtle ones, like idempotent methods:
+
+```
+Methods can also have the property of "idempotence" in that
+(aside from error or expiration issues) the side-effects of
+N > 0 identical requests is the same as for a single request.
+
+The methods GET, HEAD, PUT and DELETE share this property.
+Also, the methods OPTIONS and TRACE SHOULD NOT have side effects,
+and so are inherently idempotent.
+
+However, it is possible that a sequence of several requests
+is non- idempotent, even if all of the methods executed in
+that sequence are idempotent. (A sequence is idempotent if
+a single execution of the entire sequence always yields a
+result that is not changed by a reexecution of all,
+or part, of that sequence.) For example,
```

Comment:
that reminds me of another issue with REST over HTTP. Non-idempotent sequences of operations need to be made in the same request, and then every service that implements some kind of 'bulk operations' implements their own protocol language using XML/JSON/etc inside a POST/PUT/PATCH method and sometimes introduces HTTP status codes adding meanings as 'processed but nothing changed', 'partial success' and so on... There exist HTTP extensions to handle this cases and transactions, but there are no agreements on where it's valid or supported. In my opinion, these kinds of actions (bulks/transactions/etc) do not fit in REST nor HTTP, but people think REST model applies to everything...

Reply:
Agreed...it is sad that on the dissertation Roy Fielding is very clear that REST does not applies to everything =/